### ISSUE TITLE
General priors

### ISSUE THREAD

**Issue by martin-ueding** [HISKP-LQCD/hadron#136]
In Pull Request #88 there now is a model which has arbitrary many summands that each consist of the regular “single” model. The energies of the pollution terms are fit parameters, and currently they are unconstrained. This likely is unstable, therefore we have to implement priors to constrain them to energies extracted from one- and two-particle functions. Still I would prefer if we could land PR #88 before starting work on this such that it does not become an eternal feature branch.

The prior that constrain fit parameter `p_n` around a known value `p_n^0` with uncertainty `\Delta p_n^0` has the form `(\frac{p_n - p_n^0}{\Delta p_n^0})^2`.

The priors are additions to the χ², they are not part of the model. Therefore I do not want to have them in the model but the `bootstrap.nlsfit` function. I want an additional parameter to that function like

```r
priors = list(list(par = <integer>, val = <numeric>, boot = <numeric vector>))
```

This way one can restrict arbitrary fit parameters around a known value. This parameter must be passed down from `new_matrixfit` to `bootstrap.nlsfit`.

Implementing this for the `optim` algorithm which minimizes the χ², we can just add another summand to that and adjust the jacobian `\partial χ² / \partial par` accordingly. But as we have implemented this method in terms of the other method, we actually just have to extend the second method and get this one for free.

For `nls.lm`, which minimizes `\vec \chi`, we need to think a bit more. For a general covariance matrix `C` we take its Cholesky decomposition into `L^\dagger L` and then invert `L` to give the square root of the weight matrix, which I will call `W`. We can then effectively assemble the χ² by computing `r^\dagger W^\dagger W r` where the residual `r` is `y - f(t, \lambda)`. We pass `\chi = W r` to the `nls.lm` function.

I think that we should just extend the prediction `f` by one element for each prior with `p_n - p_n^0` and then extend the weight matrix to `W \oplus 1/(\Delta p_n)`. The target values for the independent variable, `y`, will be extended with `p_n^0`. This way we can just let it minimize out `\chi`.

At the moment I see these tasks, but that can change as implementation progresses.

- [x] Add a new parameter `priors` to `bootstrap.nlsfit`.
- [ ] Extend the various `fitchi`, `dfitchi` to also include all the terms for the given priors. There are eight functions in total (correlated/uncorrelated, prediction/jacobian, y-errors/xy-errors), so one should think about a way to minimize code duplication.
- [ ] Automatically choose the `p_n^0` as the starting value for the `n`-th fit parameter.

**Comment by kostrzewa** [comment]
> Note that even with your proposal of always using a diagonal covariance matrix for the priors, the resulting fit parameters will have the correct correlation.

actually, will they?

**Comment by martin-ueding** [comment]
In the case of three summands for a fit we would specify say *E₁* and *E₂* as priors and have *E₃* as a fit parameter. We would also have a cov(*E₁*, *E₂*) ≠ 0 from the bootstrap samples. Using a diagonal covariance matrix for the prior part in *W* will then neglect this correlation. So perhaps we should at least make *W* block diagonal with the data as one block and all the priors as another block. But I fear that one has to find an argument for getting away with block-diagonality only and the general case actually is a dense *W*. I'll have to think about this for a while.

**Comment by kostrzewa** [comment]
The covariance matrix for this should be block-diagonal. The uppper-left block of the matrix, say, contains the covariance of the data (correlation function in some range of time slices) and the lower-right block the covariance of the priors. In the case that the priors are uncorrelated, these are just diagonal entries (1/Deltap_n)^2. In the case case that the priors are correlated, this block is dense and symmetric as usual.

There is also the very general case in which the priors are correlated with the data. In that case this correlation would just have to be established. In particular, if the priors come from actual fits as will be the case here, establishing their correlation is as simple as adding them to the data matrix and computing the covariance matrix.

**Comment by martin-ueding** [comment]
I think you wrote that before our discussion, right?

We concluded that basically one just has to do `cov(cbind(data_bsamples, prior_bsamples))` and gets the full general covariance matrix for all the variables that enter the chi.
